File: binding/src/libqmovie.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmovie.html
pub const qmovie = struct {

    /// New constructs a new QMovie object.
    pub fn New() QtC.QMovie {
        return qtc.QMovie_new();
    }


    /// New2 constructs a new QMovie object.
    pub fn New2(device: ?*anyopaque) QtC.QMovie {
        return qtc.QMovie_new2(@ptrCast(device));
    }


    /// New3 constructs a new QMovie object.
    pub fn New3(fileName: []const u8) QtC.QMovie {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QMovie_new3(fileName_str);
    }


    /// New4 constructs a new QMovie object.
    pub fn New4(parent: ?*anyopaque) QtC.QMovie {
        return qtc.QMovie_new4(@ptrCast(parent));
    }


    /// New5 constructs a new QMovie object.
    pub fn New5(device: ?*anyopaque, format: []u8) QtC.QMovie {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
return qtc.QMovie_new5(@ptrCast(device), format_str);
    }


    /// New6 constructs a new QMovie object.
    pub fn New6(device: ?*anyopaque, format: []u8, parent: ?*anyopaque) QtC.QMovie {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
return qtc.QMovie_new6(@ptrCast(device), format_str, @ptrCast(parent));
    }


    /// New7 constructs a new QMovie object.
    pub fn New7(fileName: []const u8, format: []u8) QtC.QMovie {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
return qtc.QMovie_new7(fileName_str, format_str);
    }


    /// New8 constructs a new QMovie object.
    pub fn New8(fileName: []const u8, format: []u8, parent: ?*anyopaque) QtC.QMovie {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
return qtc.QMovie_new8(fileName_str, format_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMovie_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmovie.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SupportedFormats() []const u8 {
        return qtc.QMovie_SupportedFormats();
    }

    pub fn SetDevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QMovie_SetDevice(@ptrCast(self), @ptrCast(device));
    }

    pub fn Device(self: ?*anyopaque, ) QtC.QIODevice {
        return qtc.QMovie_Device(@ptrCast(self));
    }

    pub fn SetFileName(self: ?*anyopaque, fileName: []const u8) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
qtc.QMovie_SetFileName(@ptrCast(self), fileName_str);
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMovie_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmovie.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetFormat(self: ?*anyopaque, format: []u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QMovie_SetFormat(@ptrCast(self), format_str);
    }

    pub fn Format(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QMovie_Format(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qmovie.Format: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn SetBackgroundColor(self: ?*anyopaque, color: ?*anyopaque) void {
        qtc.QMovie_SetBackgroundColor(@ptrCast(self), @ptrCast(color));
    }

    pub fn BackgroundColor(self: ?*anyopaque, ) QtC.QColor {
        return qtc.QMovie_BackgroundColor(@ptrCast(self));
    }

    pub fn State(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_State(@ptrCast(self));
    }

    pub fn FrameRect(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QMovie_FrameRect(@ptrCast(self));
    }

    pub fn CurrentImage(self: ?*anyopaque, ) QtC.QImage {
        return qtc.QMovie_CurrentImage(@ptrCast(self));
    }

    pub fn CurrentPixmap(self: ?*anyopaque, ) QtC.QPixmap {
        return qtc.QMovie_CurrentPixmap(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QMovie_IsValid(@ptrCast(self));
    }

    pub fn LastError(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_LastError(@ptrCast(self));
    }

    pub fn LastErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMovie_LastErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmovie.LastErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn JumpToFrame(self: ?*anyopaque, frameNumber: i32) bool {
        return qtc.QMovie_JumpToFrame(@ptrCast(self), frameNumber);
    }

    pub fn LoopCount(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_LoopCount(@ptrCast(self));
    }

    pub fn FrameCount(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_FrameCount(@ptrCast(self));
    }

    pub fn NextFrameDelay(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_NextFrameDelay(@ptrCast(self));
    }

    pub fn CurrentFrameNumber(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_CurrentFrameNumber(@ptrCast(self));
    }

    pub fn Speed(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_Speed(@ptrCast(self));
    }

    pub fn ScaledSize(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMovie_ScaledSize(@ptrCast(self));
    }

    pub fn SetScaledSize(self: ?*anyopaque, size: ?*anyopaque) void {
        qtc.QMovie_SetScaledSize(@ptrCast(self), @ptrCast(size));
    }

    pub fn CacheMode(self: ?*anyopaque, ) i32 {
        return qtc.QMovie_CacheMode(@ptrCast(self));
    }

    pub fn SetCacheMode(self: ?*anyopaque, mode: i32) void {
        qtc.QMovie_SetCacheMode(@ptrCast(self), @intCast(mode));
    }

    pub fn Started(self: ?*anyopaque, ) void {
        qtc.QMovie_Started(@ptrCast(self));
    }

    pub fn OnStarted(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QMovie_Connect_Started(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Resized(self: ?*anyopaque, size: ?*anyopaque) void {
        qtc.QMovie_Resized(@ptrCast(self), @ptrCast(size));
    }

    pub fn OnResized(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMovie_Connect_Resized(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Updated(self: ?*anyopaque, rect: ?*anyopaque) void {
        qtc.QMovie_Updated(@ptrCast(self), @ptrCast(rect));
    }

    pub fn OnUpdated(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMovie_Connect_Updated(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn StateChanged(self: ?*anyopaque, state: i32) void {
        qtc.QMovie_StateChanged(@ptrCast(self), @intCast(state));
    }

    pub fn OnStateChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QMovie_Connect_StateChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Error(self: ?*anyopaque, errorVal: i32) void {
        qtc.QMovie_Error(@ptrCast(self), @intCast(errorVal));
    }

    pub fn OnError(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QMovie_Connect_Error(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Finished(self: ?*anyopaque, ) void {
        qtc.QMovie_Finished(@ptrCast(self));
    }

    pub fn OnFinished(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QMovie_Connect_Finished(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn FrameChanged(self: ?*anyopaque, frameNumber: i32) void {
        qtc.QMovie_FrameChanged(@ptrCast(self), frameNumber);
    }

    pub fn OnFrameChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QMovie_Connect_FrameChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Start(self: ?*anyopaque, ) void {
        qtc.QMovie_Start(@ptrCast(self));
    }

    pub fn JumpToNextFrame(self: ?*anyopaque, ) bool {
        return qtc.QMovie_JumpToNextFrame(@ptrCast(self));
    }

    pub fn SetPaused(self: ?*anyopaque, paused: bool) void {
        qtc.QMovie_SetPaused(@ptrCast(self), paused);
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QMovie_Stop(@ptrCast(self));
    }

    pub fn SetSpeed(self: ?*anyopaque, percentSpeed: i32) void {
        qtc.QMovie_SetSpeed(@ptrCast(self), percentSpeed);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMovie_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmovie.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMovie_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmovie.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMovie_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/movie.html#types
pub const enums = struct {
    pub const MovieState = enum {
        pub const NotRunning: i32 = 0;
        pub const Paused: i32 = 1;
        pub const Running: i32 = 2;
    };

    pub const CacheMode = enum {
        pub const CacheNone: i32 = 0;
        pub const CacheAll: i32 = 1;
    };

};

File: binding/src/libqmovie.cpp
#include <QByteArray>
#include <QColor>
#include <QIODevice>
#include <QImage>
#include <QMovie>
#include <QObject>
#include <QPixmap>
#include <QRect>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qmovie.h>
#include "libqmovie.h"
#include "libqmovie.hxx"

QMovie* QMovie_new() {
	 return new QMovie();
}

QMovie* QMovie_new2(QIODevice* device) {
	 return new QMovie(device);
}

QMovie* QMovie_new3(const libqt_string fileName) {
	 return new QMovie(QString::fromUtf8(fileName.data, fileName.len));
}

QMovie* QMovie_new4(QObject* parent) {
	 return new QMovie(parent);
}

QMovie* QMovie_new5(QIODevice* device, const libqt_string format) {
	 return new QMovie(device, QByteArray(format.data, format.len));
}

QMovie* QMovie_new6(QIODevice* device, const libqt_string format, QObject* parent) {
	 return new QMovie(device, QByteArray(format.data, format.len), parent);
}

QMovie* QMovie_new7(const libqt_string fileName, const libqt_string format) {
	 return new QMovie(QString::fromUtf8(fileName.data, fileName.len), QByteArray(format.data, format.len));
}

QMovie* QMovie_new8(const libqt_string fileName, const libqt_string format, QObject* parent) {
	 return new QMovie(QString::fromUtf8(fileName.data, fileName.len), QByteArray(format.data, format.len), parent);
}

libqt_string QMovie_Tr(const char* s) {
	QString _ret = QMovie::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QMovie_SupportedFormats() {
	auto _ret = QMovie::supportedFormats();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QMovie_SetDevice(QMovie* self, QIODevice* device) {
	self->setDevice(device);
}

QIODevice* QMovie_Device(const QMovie* self) {
	return self->device();
}

void QMovie_SetFileName(QMovie* self, const libqt_string fileName) {
	self->setFileName(QString::fromUtf8(fileName.data, fileName.len));
}

libqt_string QMovie_FileName(const QMovie* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMovie_SetFormat(QMovie* self, const libqt_string format) {
	self->setFormat(QByteArray(format.data, format.len));
}

libqt_string QMovie_Format(const QMovie* self) {
	QByteArray _qb = self->format();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMovie_SetBackgroundColor(QMovie* self, const QColor* color) {
	self->setBackgroundColor(*color);
}

QColor* QMovie_BackgroundColor(const QMovie* self) {
	return new QColor(self->backgroundColor());
}

int QMovie_State(const QMovie* self) {
	return self->state();
}

QRect* QMovie_FrameRect(const QMovie* self) {
	return new QRect(self->frameRect());
}

QImage* QMovie_CurrentImage(const QMovie* self) {
	return new QImage(self->currentImage());
}

QPixmap* QMovie_CurrentPixmap(const QMovie* self) {
	return new QPixmap(self->currentPixmap());
}

bool QMovie_IsValid(const QMovie* self) {
	return self->isValid();
}

int QMovie_LastError(const QMovie* self) {
	return self->lastError();
}

libqt_string QMovie_LastErrorString(const QMovie* self) {
	QString _ret = self->lastErrorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QMovie_JumpToFrame(QMovie* self, int frameNumber) {
	return self->jumpToFrame(frameNumber);
}

int QMovie_LoopCount(const QMovie* self) {
	return self->loopCount();
}

int QMovie_FrameCount(const QMovie* self) {
	return self->frameCount();
}

int QMovie_NextFrameDelay(const QMovie* self) {
	return self->nextFrameDelay();
}

int QMovie_CurrentFrameNumber(const QMovie* self) {
	return self->currentFrameNumber();
}

int QMovie_Speed(const QMovie* self) {
	return self->speed();
}

QSize* QMovie_ScaledSize(QMovie* self) {
	return new QSize(self->scaledSize());
}

void QMovie_SetScaledSize(QMovie* self, const QSize* size) {
	self->setScaledSize(*size);
}

int QMovie_CacheMode(const QMovie* self) {
	return self->cacheMode();
}

void QMovie_SetCacheMode(QMovie* self, int mode) {
	self->setCacheMode(static_cast<QFlags<QGraphicsView::CacheModeFlag>>(mode));
}

void QMovie_Started(QMovie* self) {
	self->started();
}

void QMovie_Connect_Started(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*) = reinterpret_cast<void (*)(QMovie*)>(slot);
    QMovie::connect(self, &QMovie::started, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QMovie_Resized(QMovie* self, const QSize* size) {
	self->resized(*size);
}

void QMovie_Connect_Resized(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*, const QSize*) = reinterpret_cast<void (*)(QMovie*, const QSize*)>(slot);
    QMovie::connect(self, &QMovie::resized, [self, slotFunc](const QSize& size) {
	slotFunc(self, size);
    });
}

void QMovie_Updated(QMovie* self, const QRect* rect) {
	self->updated(*rect);
}

void QMovie_Connect_Updated(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*, const QRect*) = reinterpret_cast<void (*)(QMovie*, const QRect*)>(slot);
    QMovie::connect(self, &QMovie::updated, [self, slotFunc](const QRect& rect) {
	slotFunc(self, rect);
    });
}

void QMovie_StateChanged(QMovie* self, int state) {
	self->stateChanged(static_cast<QMovie::MovieState>(state));
}

void QMovie_Connect_StateChanged(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*, int) = reinterpret_cast<void (*)(QMovie*, int)>(slot);
    QMovie::connect(self, &QMovie::stateChanged, [self, slotFunc](QMovie::MovieState state) {
	slotFunc(self, state);
    });
}

void QMovie_Error(QMovie* self, int errorVal) {
	self->error(static_cast<QImageReader::ImageReaderError>(errorVal));
}

void QMovie_Connect_Error(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*, int) = reinterpret_cast<void (*)(QMovie*, int)>(slot);
    QMovie::connect(self, &QMovie::error, [self, slotFunc](QImageReader::ImageReaderError errorVal) {
	slotFunc(self, errorVal);
    });
}

void QMovie_Finished(QMovie* self) {
	self->finished();
}

void QMovie_Connect_Finished(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*) = reinterpret_cast<void (*)(QMovie*)>(slot);
    QMovie::connect(self, &QMovie::finished, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QMovie_FrameChanged(QMovie* self, int frameNumber) {
	self->frameChanged(frameNumber);
}

void QMovie_Connect_FrameChanged(QMovie* self, intptr_t slot) {
    void (*slotFunc)(QMovie*, int) = reinterpret_cast<void (*)(QMovie*, int)>(slot);
    QMovie::connect(self, &QMovie::frameChanged, [self, slotFunc](int frameNumber) {
	slotFunc(self, frameNumber);
    });
}

void QMovie_Start(QMovie* self) {
	self->start();
}

bool QMovie_JumpToNextFrame(QMovie* self) {
	return self->jumpToNextFrame();
}

void QMovie_SetPaused(QMovie* self, bool paused) {
	self->setPaused(paused);
}

void QMovie_Stop(QMovie* self) {
	self->stop();
}

void QMovie_SetSpeed(QMovie* self, int percentSpeed) {
	self->setSpeed(percentSpeed);
}

libqt_string QMovie_Tr2(const char* s, const char* c) {
	QString _ret = QMovie::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMovie_Tr3(const char* s, const char* c, int n) {
	QString _ret = QMovie::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMovie_Delete(QMovie* self) {
    delete self;
}

